src/backend/filter.rs

```diff
@@ -0,0 +1,454 @@
+
+/// Filter containing rules assigned to syscall numbers.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SeccompFilter {
+    /// Map of syscall numbers and corresponding rule chains.
+    rules: BTreeMap<i64, Vec<SeccompRule>>,
+    /// Default action to apply to syscalls that do not match the filter.
+    default_action: SeccompAction,
+    /// Filter action to apply to syscalls that match the filter.
+    filter_action: SeccompAction,
+    /// Target architecture of the generated BPF filter.
+    target_arch: TargetArch,
+}
+
+impl SeccompFilter {
+    /// Creates a new filter with a set of rules, an on-match and default action.
+    ///
+    /// # Arguments
+    ///
+    /// * `rules` - Map containing syscall numbers and their respective [`SeccompRule`]s.
+    /// * `default_action` - [`SeccompAction`] taken for all syscalls that do not match any rule.
+    /// * `filter_action` - [`SeccompAction`] taken for system calls that match the filter.
+    /// * `target_arch` - Target architecture of the generated BPF filter.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use std::convert::TryInto;
+    /// use seccompiler::{
+    ///     SeccompAction, SeccompCmpArgLen, SeccompCmpOp, SeccompCondition, SeccompFilter,
+    ///     SeccompRule,
+    /// };
+    ///
+    /// let filter = SeccompFilter::new(
+    ///     vec![
+    ///         (libc::SYS_accept4, vec![]),
+    ///         (
+    ///             libc::SYS_fcntl,
+    ///             vec![
+    ///                 SeccompRule::new(vec![
+    ///                     SeccompCondition::new(
+    ///                         1,
+    ///                         SeccompCmpArgLen::Dword,
+    ///                         SeccompCmpOp::Eq,
+    ///                         libc::F_SETFD as u64,
+    ///                     ).unwrap(),
+    ///                     SeccompCondition::new(
+    ///                         2,
+    ///                         SeccompCmpArgLen::Dword,
+    ///                         SeccompCmpOp::Eq,
+    ///                         libc::FD_CLOEXEC as u64,
+    ///                     ).unwrap(),
+    ///                 ]).unwrap(),
+    ///                 SeccompRule::new(vec![SeccompCondition::new(
+    ///                     1,
+    ///                     SeccompCmpArgLen::Dword,
+    ///                     SeccompCmpOp::Eq,
+    ///                     libc::F_GETFD as u64,
+    ///                 ).unwrap()]).unwrap(),
+    ///             ],
+    ///         ),
+    ///     ]
+    ///     .into_iter()
+    ///     .collect(),
+    ///     SeccompAction::Trap,
+    ///     SeccompAction::Allow,
+    ///     std::env::consts::ARCH.try_into().unwrap(),
+    /// );
+    /// ```
+    ///
+    /// [`SeccompRule`]: struct.SeccompRule.html
+    /// [`SeccompAction`]: struct.SeccompAction.html
+    pub fn new(
+        rules: BTreeMap<i64, Vec<SeccompRule>>,
+        default_action: SeccompAction,
+        filter_action: SeccompAction,
```

Comment:
The value of renaming the actions comes from renaming both of them so that it is intuitive in respect to where are they applies (on match or on mis-match). I prefer us renaming both of them.

What other seccomp libraries have you looked at out of curiosity?

Reply:
libseccomp & libseccomp golang use `default_action` and `action`.
minijail does not even allow the user to specify the actions. It uses `allow` for on-match and kill or trap for mismatch (depending on what is available on the host)

Reply:
I'm fine either way, so I renamed the actions to `match_action` and `mismatch_action`, per your suggestion